packages/core/android/src/main/kotlin/com/datadog/reactnative/DdSdkBridgeExt.kt

```diff
@@ -6,44 +6,187 @@
+                    createWritableArray
+                )
+            )
+            is Map<*, *> -> map.putMap(
+                key,
+                v.toWritableMap(
+                    createWritableMap,
+                    createWritableArray
+                )
+            )
+            else -> Log.e(
+                javaClass.simpleName,
+                "Unhandled type ${v.javaClass.simpleName} has been ignored"
+            )
         }
     }
+
     return map
 }
+
+/**
+ * Recursively converts the [ReadableMap] to a [Map] which only contains Kotlin stdlib objects,
+ * such as [List], [Map] and the raw types.
+ */
+internal fun ReadableMap.toMap(): Map<String, Any> {
+    val map: HashMap<String, Any> = this.toHashMap()
+    val iterator: Iterator<String> = map.keys.iterator()
+
+    while (iterator.hasNext()) {
+        val key = iterator.next()
+        try {
+            when (val type = getType(key)) {
+                ReadableType.Map -> {
+                    map[key] = getMap(key)?.toMap() ?: Log.e(
+                        javaClass.simpleName,
+                        "Cannot convert nested map for key: $key"
+                    )
+                }
+
+                ReadableType.Array -> {
+                    map[key] = getArray(key)?.toList() ?: Log.e(
+                        javaClass.simpleName,
+                        "Cannot convert nested array for key: $key"
+                    )
+                }
+
+                ReadableType.Null, ReadableType.Boolean, ReadableType.Number, ReadableType.String -> {}
+                else -> Log.e(
+                    javaClass.simpleName,
+                    "Skipping unhandled type [${type.name}] for key: $key"
+                )
+            }
+        } catch (err: IllegalArgumentException) {
+            Log.e(javaClass.simpleName, "Could not convert object for key: $key.", err)
+        }
+    }
+
+    return map
+}
+
+/**
+ * Recursively converts the [ReadableArray] to a [List] which only contains Kotlin stdlib objects,
+ * such as [List], [Map] and the raw types.
+ * or [List], instead of [ReadableMap] and [ReadableArray] respectively).
+ */
+internal fun ReadableArray.toList(): List<*> {
+    val out = mutableListOf<Any?>()
+    for (i in 0 until size()) {
+        // ReadableArray throws a null pointer exception if getMap(i) or getArray(i) returns null
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            when (val type = getType(i)) {
+                ReadableType.Null -> out.add(null)
+                ReadableType.Boolean -> out.add(getBoolean(i))
+                ReadableType.Number -> out.add(getDouble(i))
+                ReadableType.String -> out.add(getString(i))
+                ReadableType.Map -> out.add(getMap(i).toMap())
+                ReadableType.Array -> out.add(getArray(i).toList())
+                else -> Log.e(javaClass.simpleName, "Unhandled ReadableType: ${type.name}.")
+            }
```

Comment:
a bit less code and non-nullable list

```suggestion
            val value = when (val type = getType(i)) {
                ReadableType.Null -> null
                ReadableType.Boolean -> getBoolean(i)
                ReadableType.Number -> getDouble(i)
                ReadableType.String -> getString(i)
                ReadableType.Map -> getMap(i).toMap()
                ReadableType.Array -> getArray(i).toList()
                else -> Log.e(javaClass.simpleName, "Unhandled ReadableType: ${type.name}.")
            }
            if (value != null) list += value
```